interfaces/builtin/pluggable_storage.go

```diff
@@ -0,0 +1,71 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+const pluggableStorageConnectedPlugAppArmor = `
+# Description: Can mount and unmount removable storage. This is restricted
+# because it gives privileged access to mount commands and should only be used
+# with trusted apps.
+# Usage: reserved
+
+# Needed for mount/unmount operations
+capability sys_admin,
+
+# Mount/unmount USB storage devices
+mount /dev/sd* -> /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+umount /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+
+# Mount/unmount SD cards
+mount /dev/mmcblk* -> /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+umount /var/snap/${SNAP_NAME}/${SNAP_REVISION}/**,
+
+# Allow calling the system mount/umount binaries to do the dirty work
+/bin/mount ixr,
+/bin/umount ixr,
+
+# mount/umount (via libmount) track some mount info in these files
+/run/mount/utab wrl,
+
+# USB dev files (sda, sdb, etc...)
+/dev/sd* r,
+
+# SD card dev files
+/dev/mmcblk* r,
```

Comment:
This interface appears to intend (there is no docs/interfaces.md update and the Description is too terse so just going by the policy) to allow readonly access to pluggable devices but it actually grants ownership of the device.

The mount rules for /dev/sd* mean that the device can be mounted without mount rules, which defaults to read/write. As such, a snap with this interface connected can simply: `mount /dev/sda1 $SNAP_DATA/mnt` and on a typical install modify all files in the sda1 partition-- /etc/shadow, sshd_config, pam, grub, security policy, systemd jobs, etc, etc. On an all snaps system this interface could be used to replace security policy, systemd jobs, etc, etc. Once that is done, it is trivial to completely own the device on reboot (and a boot can be retriggered easily enough with what this interface grants). The same can be said for devices the boot off of SD cards like beaglebone or rpi due to the mount rule for /dev/mmcblk*. We could adjust the mount options in the apparmor rule to be readonly but I suspect that will not provide the required functionality that this interface intends to provide.

Furthermore, the /dev/sd* and /dev/mmcblk* read access means that this interface grants complete read access to these raw devices and partitions (and therefore the entire device) and that sensitive system, user and application data is granted to any snap with this interface connected.

The sys_admin capability is unfortunate but required due to the use of mount, and due to limitation in the Linux capabilities system, this grants more than simply mount.

In all, -1 for this interface as implemented.

This needs quite a bit of design to work well and safely. I think there are at least three options for a proper solution that would need to be considered:

1. use something along the lines of the storage framework as designed by @vosst for Ubuntu Touch
2. snapd should grow a device monitor service that is able to detect pluggable device events and make the the filesystem available to snaps that have this interface connected
3. snapd should grow a device monitor service that is able to detect pluggable device events and make them available to snaps that have this interface connected via the SecurityUDev backend. Ie, tag assigned devices so they are available to snaps.

'1' would need @vosst to comment and review for snappy (IIRC it was designed only with Touch in mind, but it may be adaptable to snapd)

'2' is a reimplementation of '1' AIUI in a snappy world and does not require giving the snap access to `mount`. Instead, the device monitor mounts and the security policy of the snap is adjusted accordingly (and perhaps we would simply use a bind mount into /var/lib/$SNAP_NAME/mounts so the security policy doesn't have to be adjusted).

'3' would still require the mount command, but we are able to limit what can be mounted to only those devices that are in the device cgroup, so on a device with /dev/sda for the SSD/hardrive, /dev/sda1 wouldn't be there (since the device monitor wouldn't add it) but /dev/sdb is. Specifically, the device monitor would update the udev rules for any plugged in removable block device for the snap, which would trigger the udev command to add the devices to the snap's device cgroup.

In summary-- we don't want the snap to have to run mount commands on random devices in /dev itself. Instead, something else (ie, a trusted helper outside of the snap's process space) is in charge of mounting the devices and making the filesystem (not the raw device) available to the snap. This needs proper design with the interfaces team along with other members of the security team (eg, @tyhicks and/or Emily) and likely @vosst so that whatever is done works with converged devices.